src/protego.py

```diff
@@ -0,0 +1,273 @@
+RequestRate = namedtuple(
+    'RequestRate', ['requests', 'seconds', 'start_time', 'end_time'])
+
+DISALLOW_DIRECTIVE = {'disallow', 'dissallow', 'dissalow', 'disalow', 'diasllow', 'disallaw'}
+ALLOW_DIRECTIVE = {'allow'}
+USER_AGENT_DIRECTIVE = {'user-agent', 'useragent', 'user agent'}
+SITEMAP_DIRECTIVE = {'sitemap', 'sitemaps', 'site-map'}
+CRAWL_DELAY_DIRECTIVE = {'crawl-delay', 'crawl delay'}
+REQUEST_RATE_DIRECTIVE = {'request-rate', 'request rate'}
+
+
+class RuleSet:
+
+    def __init__(self):
+        self.user_agent = None
+        self.rules = []
+        self.crawl_delay = None
+        self.rules_sorted = False
+        self.req_rate = None
+
+    def set_user_agent(self, user_agent):
+        self.user_agent = user_agent
+
+    def applies_to(self, robotname):
+        robotname = robotname.strip().lower()
+        if self.user_agent == '*':
+            return 1
+        if self.user_agent in robotname:
+            return len(self.user_agent)
+        return 0
+
+    def _quote_path(self, path, safe='/'):
+        if path == '/?' or path == '/;':
+            return path
+        parts = urlparse(path)
+        path = re.sub("%2[fF]", "\n", parts.path)
+        if six.PY2:
+            path = quote(unquote(path.encode('utf-8')), safe=safe).decode('utf-8')
+        else:
+            path = quote(unquote(path), safe=safe)
+        path.replace("\n", "%2F")
+        parts = ParseResult('', '', path, parts.params, parts.query, parts.fragment)
+        path = urlunparse(parts)
+        return path
+
+    def allow(self, path):
+        path = self._quote_path(path, safe='/*$')
+        self.rules.append(Rule(field='allow', value=path))
+
+        # if index.html is allowed, we interpret this as / being allowed too.
+        if path.endswith('/index.html'):
+            self.allow(path[:-10] + '$')
+
+    def disallow(self, path):
+        path = self._quote_path(path, safe='/*$')
+        self.rules.append(Rule(field='disallow', value=path))
+
+    def _parser_match(self, pattern, url):
+        pattern = re.sub(r'\*+', '*', pattern)
+        s = re.split(r'([$*])', pattern)
+        for index, substr in enumerate(s):
+            if substr not in ['*', '$']:
+                s[index] = re.escape(substr)
+            else:
+                s[index] = s[index].replace('*', '.*')
+        pattern = '^' + ''.join(s)
+        if re.findall(pattern, url):
+            return True
+        return False
+
+    def allowed(self, url):
+        if not self.rules_sorted:
+            self.rules = sorted(self.rules, key=lambda r: (len(r.value), r.field == 'allow'), reverse=True)
+            self.rules_sorted = True
+        url = self._quote_path(url)
+        allowed = True
+        for rule in self.rules:
+            if self._parser_match(rule.value, url):
+                if rule.field == 'disallow':
+                    allowed = False
+                break
+        return allowed
+
+    def set_crawl_delay(self, delay):
```

Comment:
Consider using a `@property` instead of a getter and a setter. Also applies to `user_agent` and `request_rate`.